### ISSUE TITLE
Agent won't start due to issues with HW TPM

### ISSUE THREAD

**Issue by kkaarreell** [keylime/rust-keylime#438]
# Environment
 * OS / version: CentOS Stream 9
 * Processor architecture:  x86_64
 * TPM Manufacturer: TPM on Intel® Server S2600 motherboard
 * Keylime version: current upstream, keylime 6f374d5 , rust-keylime 0186093

# Description 

I am not able to make keylime Rust agent work with HW TPM. Everything works well with SW TPM or even for HW TPM when Python agent is used.

## Expected behavior vs. actual behavior 
rust agent won't register to registrat

## Steps to reproduce problem

I have tested two initial setups:

A) Running tpm2_clear following the test 
https://github.com/RedHat-SP-Security/keylime-tests/blob/main/functional/basic-attestation-on-localhost/test.sh
No tpm_ownerpassword is configured in keylime-agent.conf

With this setup I am not able to register agent to registrar. The registrar log says:
```
2022-08-16 06:19:24.268 - keylime.registrar_client - WARNING - TLS is enabled.
2022-08-16 06:19:24.268 - keylime.registrar_client - INFO - Setting up client TLS...
2022-08-16 06:19:24.279 - keylime.registrar_client - CRITICAL - Error: could not get agent d432fbb3-d2f1-4a97-9ef7-75bd81c00000 data from Registrar Server: 404
2022-08-16 06:19:24.279 - keylime.registrar_client - CRITICAL - Response code 404: agent d432fbb3-d2f1-4a97-9ef7-75bd81c00000 not found
```

Later, in agent.log there is nothing useful
```
Uploading var-tmp-limeLib-registrar.log .done
Aug 16 06:18:41 systemd[1]: Started The Keylime compute agent.
Aug 16 06:18:41 keylime_agent[25444]:  WARN  keylime_agent::common > Agent Data not available under: /var/lib/keylime/agent_data.json
Aug 16 06:18:41 keylime_agent[25444]:  WARN  keylime_agent::common > Cannot find value for key revocation_actions in file /etc/keylime-agent.conf
Aug 16 06:18:41 keylime_agent[25444]:  WARN  keylime_agent::common > Cannot find value for key run_as in file /etc/keylime-agent.conf
Aug 16 06:18:41 keylime_agent[25444]:  WARN  keylime_agent::common > Cannot drop privileges since 'run_as' is empty in 'cloud_agent' section of keylime-agent.conf.
Aug 16 06:18:41 keylime_agent[25444]:  DEBUG keylime_agent::secure_mount > Secure store location /var/lib/keylime/secure not mounted
Aug 16 06:18:41 keylime_agent[25444]:  INFO  keylime_agent::secure_mount > Directory "/var/lib/keylime/secure" created.
Aug 16 06:18:41 keylime_agent[25444]:  INFO  keylime_agent::secure_mount > Mounting secure storage location "/var/lib/keylime/secure" on tmpfs.
Aug 16 06:18:41 keylime_agent[25444]:  INFO  keylime_agent               > Starting server with API version v2.0...
Aug 16 06:21:45 systemd[1]: Stopping The Keylime compute agent...
Aug 16 06:21:45 systemd[1]: keylime_agent.service: Deactivated successfully.
Aug 16 06:21:45 systemd[1]: Stopped The Keylime compute agent.
```

B) The second scenario starts with `tpm2_clear` and taking the ownership
```
# tpm2_clear
# tpm2_changeauth -c o keylime
# tpm2_changeauth -c e keylime
```
I configure `tpm_ownerpassword = keylime` password in kelime-agent.conf.
This time agent log contains the following error:
```
keylime_agent[8947]:  INFO  keylime_agent              > Starting server with API version v2.0...
keylime_agent[8947]: WARNING:esys:src/tss2-esys/api/Esys_CreatePrimary.c:393:Esys_CreatePrimary_Finish() Received TPM Error
keylime_agent[8947]: ERROR:esys:src/tss2-esys/api/Esys_CreatePrimary.c:135:Esys_CreatePrimary() Esys Finish ErrorCode (0x000009a2)
keylime_agent[8947]: Error: Tpm { err: Tss2Error(FormatOne(FormatOneResponseCode { .0: 2466, error_number: 34, parameter: false, format_selector: true, number: 9 })), kind: Some(BadAuth), message: "authorization failure without DA implications (associated with session number 1)" }
systemd[1]: keylime_agent.service: Main process exited, code=exited, status=1/FAILURE
```

**Comment by kkaarreell** [comment]
Unfortunately, I cannot test it with Fedora since my test system is configured in a way that allows only provisioning of RHEL.
Today, I tried to create EEK manually and configured agent with `tpm_ownerpassword` and `ek_handle`.

```
# tpm2_createek -P keylime -w keylime -c 0x81010001 -G rsa
# tpm2_getcap handles-persistent
- 0x81010001
```
After that, test fails again but this time there is not useful message in agent.log
```
Aug 18 07:11:39 systemd[1]: Started The Keylime compute agent.
Aug 18 07:11:39 keylime_agent[52369]:  WARN  keylime_agent::common > Agent Data not available under: /var/lib/keylime/agent_data.json
Aug 18 07:11:39 keylime_agent[52369]:  WARN  keylime_agent::common > Cannot find value for key revocation_actions in file /etc/keylime-agent.conf
Aug 18 07:11:39 keylime_agent[52369]:  WARN  keylime_agent::common > Cannot find value for key run_as in file /etc/keylime-agent.conf
Aug 18 07:11:39 keylime_agent[52369]:  WARN  keylime_agent::common > Cannot drop privileges since 'run_as' is empty in 'cloud_agent' section of keylime-agent.conf.
Aug 18 07:11:39 keylime_agent[52369]:  DEBUG keylime_agent::secure_mount > Secure store location /var/lib/keylime/secure already mounted on tmpfs
Aug 18 07:11:39 keylime_agent[52369]:  INFO  keylime_agent               > Starting server with API version v2.0...
Aug 18 07:14:45 systemd[1]: Stopping The Keylime compute agent...
Aug 18 07:14:45 systemd[1]: keylime_agent.service: Deactivated successfully.
Aug 18 07:14:45 systemd[1]: Stopped The Keylime compute agent.
```
Moreover, after the test I am not able to access TPM
```
# tpm2_getcap handles-persistent

** (process:54756): WARNING **: 07:18:24.370: Failed to create connection with service: GDBus.Error:org.freedesktop.DBus.Error.NoReply: Remote peer disconnected
WARNING:tcti:src/tss2-tcti/tctildr.c:79:tcti_from_init() TCTI init for function 0x7f0f00e21210 failed with a0008 
WARNING:tcti:src/tss2-tcti/tctildr.c:109:tcti_from_info() Could not initialize TCTI named: tcti-abrmd 
ERROR:tcti:src/tss2-tcti/tctildr-dl.c:154:tcti_from_file() Could not initialize TCTI file: libtss2-tcti-tabrmd.so.0 
ERROR:tcti:src/tss2-tcti/tcti-device.c:501:Tss2_Tcti_Device_Init() timeout waiting for response from fd 7 
WARNING:esys:src/tss2-esys/api/Esys_GetCapability.c:303:Esys_GetCapability_Finish() Received TPM Error 
ERROR:esys:src/tss2-esys/api/Esys_GetCapability.c:107:Esys_GetCapability() Esys Finish ErrorCode (0x00000084) 
ERROR: Esys_GetCapability(0x84) - tpm:handle(unk):value is out of range or is not correct for the context
ERROR: Unable to run tpm2_getcap
```
Btw, I am not using abrmd on test system.

**Comment by kkaarreell** [comment]
Btw, I have noticed that in a console I have several messages like this
```
[  851.398261] ima: Error Communicating to TPM chip, result: 257
[  851.408613] tpm tpm0: A TPM error (257) occurred attempting extend a PCR value
```

**Comment by kkaarreell** [comment]
Another console errors appearing after 
```
tpm2_changeauth -c owner
tpm2_changeauth -c endorsement -p keylime
```
as mentioned in https://github.com/keylime/rust-keylime/issues/175#issuecomment-806941032
```
[  282.203425] tpm tpm0: tpm2_save_context: failed with a TPM error 0x0101
[  282.211886] tpm tpm0: A TPM error (257) occurred flushing context
[  282.219899] tpm tpm0: A TPM error (257) occurred flushing context
[  282.227885] tpm tpm0: A TPM error (257) occurred flushing context
[  282.233977] tpm tpm0: tpm2_commit_space: error -14
```

**Comment by kkaarreell** [comment]
And to add something important. These errors appear after I attempt to start the rust agent. There are no console errors before that and since I have kernel IMA configured, I can see the PCR10 hash changes as I execute various binaries. In other words, TPM seems to work well from this perspective.

**Comment by lkatalin** [comment]
I will see what I can find about this today.

**Comment by lkatalin** [comment]
It looks like the `257` error is broad and means the TPM has already failed:
```
[root@localhost ~]# tpm2_rc_decode 257
tpm:error(2.0): commands not being accepted because of a TPM failure
```

It looks like the Rust agent is failing before requesting registration. Otherwise you would have agent output [announcing the UUID](https://github.com/keylime/rust-keylime/blob/master/src/main.rs#L525) and that it is [requesting registration](https://github.com/keylime/rust-keylime/blob/master/src/registrar_agent.rs#L126). Based on the last output message being about the API version, I would guess the failure occurs when [getting the TPM context](https://github.com/keylime/rust-keylime/blob/master/src/main.rs#L426). This would make sense for a TPM error.

Does `TPM2_GetTestResult` output anything useful after the failure?
Does this also happen if you do not run the Python agent first on the machine, but start the Rust agent only?
Does using the Python agent after this failure work?

Also, I retract my earlier statement about Avery getting this working on hardware TPM; that setup used vTPM.

I'm still working to see if I can reproduce it.

**Comment by ansasaki** [comment]
Could it be that this is fixed by #448 ?

**Comment by kkaarreell** [comment]
Yes, I believe this issue is already outdated and resolved with https://github.com/keylime/rust-keylime/pull/448
